x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/spatial/SpatialPushDownTestCase.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * For doing that we create two indices, one fully indexed and another with index
+ * and doc values disabled. Then we index the same data in both indices and we check
+ * that the same ES|QL queries produce the same results in both.
+ */
+public abstract class SpatialPushDownTestCase extends ESIntegTestCase {
+
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(EsqlPlugin.class, SpatialPlugin.class);
+    }
+
+    /**
+     * Elasticsearch field type
+     */
+    protected abstract String fieldType();
+
+    /**
+     * A random {@link Geometry} to be indexed.
+     */
+    protected abstract Geometry getIndexGeometry();
+
+    /**
+     * A random {@link Geometry} to be used for querying.
+     */
+    protected abstract Geometry getQueryGeometry();
+
+    /**
+     * Necessary to build a ES|QL query. It should be "TO_GEOSHAPE" for geo
+     * fields and "TO_CARTESIANSHAPE" for cartesian fields.
+     */
+    protected abstract String castingFunction();
+
+    public void testPushedDownQueriesSingleValue() throws RuntimeException {
+        assertPushedDownQueries(false);
+    }
+
+    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/110830")
+    public void testPushedDownQueriesMultiValue() throws RuntimeException {
+        assertPushedDownQueries(true);
+    }
+
+    private void assertPushedDownQueries(boolean multiValue) throws RuntimeException {
+        assertAcked(prepareCreate("indexed").setMapping(String.format(Locale.ROOT, """
+            {
+              "properties" : {
+               "location": { "type" : "%s" }
+              }
+            }
+            """, fieldType())));
+
+        assertAcked(prepareCreate("not-indexed").setMapping(String.format(Locale.ROOT, """
+            {
+              "properties" : {
+               "location": { "type" : "%s",  "index" : false, "doc_values" : false }
+              }
+            }
+            """, fieldType())));
+        for (int i = 0; i < random().nextInt(50, 100); i++) {
+            if (multiValue) {
+                final String[] values = new String[randomIntBetween(1, 5)];
+                for (int j = 0; j < values.length; j++) {
+                    values[j] = "\"" + WellKnownText.toWKT(getIndexGeometry()) + "\"";
+                }
+                index("indexed", i + "", "{\"location\" : " + Arrays.toString(values) + " }");
+                index("not-indexed", i + "", "{\"location\" : " + Arrays.toString(values) + " }");
+            } else {
+                final String value = WellKnownText.toWKT(getIndexGeometry());
+                index("indexed", i + "", "{\"location\" : \"" + value + "\" }");
+                index("not-indexed", i + "", "{\"location\" : \"" + value + "\" }");
+            }
+        }
+
+        refresh("indexed", "not-indexed");
+
+        for (int i = 0; i < 10; i++) {
+            final Geometry geometry = getQueryGeometry();
```

Comment:
Since the queryGeometry and the indexGeometry are both independently random, it seems likely that they will almost always not intersect. Isn't there a risk that this could mask bugs? Perhaps we want to also test geometries we know will intersect, for example take an envelope of the query geometry?

Reply:
WE have many test like this and they have proven pretty solid so far.